django/db/models/sql/query.py

```diff
@@ -1142,6 +1143,10 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
         """
         if isinstance(filter_expr, dict):
             raise FieldError("Cannot parse keyword query as dict")
+        if isinstance(filter_expr, Expression) and isinstance(filter_expr.output_field, BooleanField):
```

Comment:
Do you think we should build the "is this a boolean field" logic as a method within the expression? It'd hide the mess a bit.

Reply:
Turns out there's a new thing called `conditional`, that I've co-opted for this.